gtsfm/utils/rotation.py

```diff
@@ -0,0 +1,38 @@
+"""Utility functions for rotations.
+
+Authors: Ayush Baid
+"""
+from typing import Dict, List, Tuple
+
+import networkx as nx
+import numpy as np
+from gtsam import Rot3
+
+
+def random_rotation() -> Rot3:
+    """Sample a random rotation by generating a sample from the 4d unit sphere."""
+    q = np.random.randn(4) * 0.03
+    # make unit-length quaternion
+    q /= np.linalg.norm(q)
+    qw, qx, qy, qz = q
+    return Rot3(qw, qx, qy, qz)
+
+
+def initialize_global_rotations_using_mst(num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Rot3]) -> List[Rot3]:
+    # Create a graph from the relative rotations dictionary
+    graph = nx.Graph()
+    for i1, i2 in i2Ri1_dict.keys():
+        # TODO: use inlier count as weight
+        graph.add_edge(i1, i2, weight=1)
+
+    # Compute the Minimum Spanning Tree (MST)
+    mst = nx.minimum_spanning_tree(graph)
+
+    wRis = [random_rotation() for _ in range(num_images)]
```

Comment:
QQ -- why initialize with random values? We can compute a true MST with real values

```
        edges = i2Ri1_dict.keys()

        # Find the largest connected component.
        cc_nodes = graph_utils.get_nodes_in_largest_connected_component(edges)
        cc_nodes = sorted(cc_nodes)

        wRi_list = [None] * num_images
        # Choose origin node.
        origin_node = cc_nodes[0]
        wRi_list[origin_node] = Rot3()

        G = nx.Graph()
        G.add_edges_from(edges)

        # Ignore 0th node, as we already set its global pose as the origin.
        for dst_node in cc_nodes[1:]:
            # Determine the path to this node from the origin. ordered from [origin_node,...,dst_node]
            path = nx.shortest_path(G, source=origin_node, target=dst_node)

            wRi = Rot3()
            for i1, i2 in zip(path[:-1], path[1:]):
                # NOTE: i1, i2 may not be in sorted order here. May need to reverse ordering.
                if i1 < i2:
                    i1Ri2 = i2Ri1_dict[(i1, i2)].inverse()
                else:
                    i1Ri2 = i2Ri1_dict[(i2, i1)]

                # wRi = wR0 * 0R1
                wRi = wRi.compose(i1Ri2)

            wRi_list[dst_node] = wRi

        return wRi_list
```

Reply:
I added a random initialization for some nodes which might not be in the MST.

Reply:
But I see that the `minimum_spanning_tree` returns a forest in that case, and I can use that.

Reply:
> QQ -- why initialize with random values? We can compute a true ST (non-MST) with real values
> 
> ```
>         edges = i2Ri1_dict.keys()
> 
>         # Find the largest connected component.
>         cc_nodes = graph_utils.get_nodes_in_largest_connected_component(edges)
>         cc_nodes = sorted(cc_nodes)
> 
>         wRi_list = [None] * num_images
>         # Choose origin node.
>         origin_node = cc_nodes[0]
>         wRi_list[origin_node] = Rot3()
> 
>         G = nx.Graph()
>         G.add_edges_from(edges)
> 
>         # Ignore 0th node, as we already set its global pose as the origin.
>         for dst_node in cc_nodes[1:]:
>             # Determine the path to this node from the origin. ordered from [origin_node,...,dst_node]
>             path = nx.shortest_path(G, source=origin_node, target=dst_node)
> 
>             wRi = Rot3()
>             for i1, i2 in zip(path[:-1], path[1:]):
>                 # NOTE: i1, i2 may not be in sorted order here. May need to reverse ordering.
>                 if i1 < i2:
>                     i1Ri2 = i2Ri1_dict[(i1, i2)].inverse()
>                 else:
>                     i1Ri2 = i2Ri1_dict[(i2, i1)]
> 
>                 # wRi = wR0 * 0R1
>                 wRi = wRi.compose(i1Ri2)
> 
>             wRi_list[dst_node] = wRi
> 
>         return wRi_list
> ```

Your suggested code will lead to `None` for camera which are not in the largest connected component. I chose to leave them as random instead.

Reply:
@ayushbaid But we prune to the largest CC before calling Shonan:

https://github.com/borglab/gtsfm/blob/master/gtsfm/multi_view_optimizer.py#L125

So I don't think we would ever encounter such a case.